gcloud-java-resourcemanager/src/main/java/com/google/gcloud/resourcemanager/testing/LocalResourceManagerHelper.java

```diff
@@ -0,0 +1,593 @@
+      return Error.PERMISSION_DENIED.response(
+          "Error when replacing " + project.getProjectId() + " because the project was not found.");
+    } else if (!oldProject.getLifecycleState().equals("ACTIVE")) {
+      return Error.FAILED_PRECONDITION.response("Error when replacing " + project.getProjectId()
+          + " because the lifecycle state was not ACTIVE.");
+    } else if (!Objects.equal(oldProject.getParent(), project.getParent())) {
+      return Error.INVALID_ARGUMENT.response(
+          "The server currently only supports setting the parent once "
+          + "and does not allow unsetting it.");
+    }
+    project.setLifecycleState("ACTIVE");
+    project.setProjectNumber(oldProject.getProjectNumber());
+    project.setCreateTime(oldProject.getCreateTime());
+    project.setParent(oldProject.getParent());
+    projects.put(project.getProjectId(), project);
+    try {
+      return new Response(HttpURLConnection.HTTP_OK, jsonFactory.toString(project));
+    } catch (IOException e) {
+      return Error.INTERNAL_ERROR.response(
+          "Error when serializing project " + project.getProjectId());
+    }
+  }
+
+  Response undelete(String projectId) {
+    Project project = projects.get(checkNotNull(projectId));
+    Response response;
+    if (project == null) {
+      // when possible, change this to 404 (#440)
+      response = Error.PERMISSION_DENIED.response(
+          "Error when undeleting " + projectId + " because the project was not found.");
+    } else if (!project.getLifecycleState().equals("DELETE_REQUESTED")) {
+      response = Error.FAILED_PRECONDITION.response("Error when undeleting " + projectId
+          + " because the lifecycle state was not DELETE_REQUESTED.");
+    } else {
+      project.setLifecycleState("ACTIVE");
+      response = new Response(HttpURLConnection.HTTP_OK, "{}");
+    }
+    return response;
+  }
+
+  private LocalResourceManagerHelper() {
+    InetSocketAddress addr = new InetSocketAddress(0);
+    try {
+      server = HttpServer.create(addr, 0);
+      port = server.getAddress().getPort();
+      server.createContext("/", new RequestHandler());
+    } catch (IOException e) {
+      throw new RuntimeException("Could not bind the mock Resource Manager server.", e);
+    }
+  }
+
+  /**
+   * Creates a LocalResourceManagerHelper object that listens to requests on the local machine.
+   */
+  public static LocalResourceManagerHelper create() {
+    return new LocalResourceManagerHelper();
+  }
+
+  /**
+   * Returns the port that the LocalResourceManagerHelper listens to for requests.
+   */
+  public int port() {
+    return port;
+  }
+
+  /**
+   * Starts the thread that runs the Resource Manager server.
+   */
+  public void start() {
+    server.start();
+  }
+
+  /**
+   * Stops the thread that runs the mock Resource Manager server.
+   */
+  public void stop() {
+    server.stop(1);
+  }
+
+  /**
+   * Utility method to add a project.
```

Comment:
I removed all utility functions except `changeLifecycleState` (important for testing when a project goes `ACTIVE` -> `DELETE_REQUESTED` -> `DELETE_IN_PROGRESS`) and `removeProject` so that a project can truly be deleted (which happens in the Cloud Resource Manager after the `DELETE_IN_PROGRESS` state)